password_security/models/res_users.py

```diff
@@ -76,19 +52,41 @@ def _check_password_rules(self, password):
         if not password:
             return True
         company_id = self.company_id
-        password_regex = ['^']
-        if company_id.password_lower:
-            password_regex.append('(?=.*?[a-z])')
-        if company_id.password_upper:
-            password_regex.append('(?=.*?[A-Z])')
-        if company_id.password_numeric:
-            password_regex.append(r'(?=.*?\d)')
-        if company_id.password_special:
-            password_regex.append(r'(?=.*?\W)')
-        password_regex.append('.{%d,}$' % company_id.password_length)
-        if not re.search(''.join(password_regex), password):
-            raise PassError(self.password_match_message())
-        return True
+        message = []
```

Comment:
IMO this code just gained complexity and reduced efficiency. Why can we not continue to use regex for everything? This is a quick regex I wrote illustrating a check for 1 uppercase, 2 lowercase, 3 digits, and 4 special characters

```
/^(?=(.*[^A-Za-z0-9]){4,})(?=(.*[A-Z]){1,})(?=(.*\d){3,})(?=.*[a-z]){2,}).+/
```

Otherwise, we iterate the string 3 times to perform evaluations at each letter, just to regex it anyways 😦

Reply:
+1 More regex, regex, regex.

Reply:
I think @smurf-U wanted to get more granular error messages, which I think is a good thing because, although you're right and possibly performance gets hit, this is really not a performance-critical part of Odoo: you're not changing passwords all the time after all.

So I'm OK on sacrificing performance for boosting UX this time.

It's also true that individual checks could use individual regexps OTOH.

Reply:
The flip side of this implementation is that there is now no way to display all of the password requirements at once. We directly put the output of the `password_match_message` in our template. IMO being able to communicate all of the rules at once is more valuable than communicating the one that failed.

Reply:
The ideal is to put the full message but highlighting any way (for example, with an ending `**`) the infringement.

Reply:
So something I was playing around with to make this work, but also not increase complexity a bunch, was to use named groups. Each character type (chars, upper, lower, numeric) would be a named group, then you simply take the length of each group and compare against what you want.

We ended up deciding meh though. I think this is the best way to go about something like that if it were to be implemented though.